### PR TITLE
Leader: Proximity Sensors

### PR DESCRIPTION
**Pull request by gabryelreyes**
- Added proximity sensors to leader and speed regulation based on distance to object in the way.
- Added V2V Emergency event
- Check V2V and RU states once. Prevent continous calling of error state

### CODE REVIEW

File: lib/ConvoyFollower/src/App.cpp [hunk 1]
```diff
@@ -227,25 +227,35 @@ void App::loop()
...
 
-    case SMPChannelPayload::STATUS_FLAG_ERROR:
-        LOG_DEBUG("RU Status ERROR.");
-        setErrorState();
-        m_statusTimeoutTimer.stop();
-        break;
+        switch (status)
+        {
+        case SMPChannelPayload::STATUS_FLAG_OK:
+            /* Nothing to do. All good. */
```
Comment (BlueAndi) [inline lib/ConvoyFollower/src/App.cpp:247]: If there is only one good case and all others are errors, you can use a simple if/else

File: lib/ConvoyFollower/src/App.cpp [hunk 2]
```diff
@@ -458,30 +468,43 @@ void App::processV2VCommunication()
...
+            /* Fallthrough */
+        case V2VCommManager::V2V_STATUS_OLD_WAYPOINT:
+            LOG_ERROR("Follower Error");
+            setErrorState();
+            break;
 
-    default:
-        break;
+        case V2VCommManager::V2V_STATUS_EMERGENCY:
+            setErrorState();
```
Comment (BlueAndi) [inline lib/ConvoyFollower/src/App.cpp:497]: What is the difference between error and emergency?

File: lib/ConvoyFollower/src/App.h [hunk 3]
```diff
@@ -78,7 +78,9 @@ class App
         m_commandTimer(),
         m_motorSpeedTimer(),
         m_statusTimer(),
-        m_statusTimeoutTimer()
+        m_statusTimeoutTimer(),
+        m_lastV2VStatus(V2VCommManager::V2VStatus::V2V_STATUS_OK),
```
Comment (BlueAndi) [inline lib/ConvoyFollower/src/App.h:82]: Shouldn't it be something like not init?

File: lib/ConvoyLeader/src/App.cpp [hunk 4]
```diff
@@ -236,25 +236,35 @@ void App::setLatestVehicleData(const Waypoint& waypoint)
...
 
-    case SMPChannelPayload::STATUS_FLAG_ERROR:
-        LOG_ERROR("RU Status ERROR.");
-        setErrorState();
-        m_statusTimeoutTimer.stop();
-        break;
+        switch (status)
+        {
+        case SMPChannelPayload::STATUS_FLAG_OK:
+            /* Nothing to do. All good. */
```
Comment (BlueAndi) [inline lib/ConvoyLeader/src/App.cpp:256]: Same point as before.

File: lib/ConvoyLeader/src/App.h [hunk 5]
```diff
@@ -79,7 +79,9 @@ class App
         m_commandTimer(),
         m_motorSpeedTimer(),
         m_statusTimer(),
-        m_statusTimeoutTimer()
+        m_statusTimeoutTimer(),
+        m_lastV2VStatus(V2VCommManager::V2VStatus::V2V_STATUS_OK),
```
Comment (BlueAndi) [inline lib/ConvoyLeader/src/App.h:83]: not init?

File: lib/ConvoyLeader/src/DrivingState.cpp [hunk 6]
```diff
@@ -73,12 +73,24 @@ void DrivingState::process(StateMachine& sm)
...
 
+        /* Limit speed based on the proximity sensors. */
+        /* m_vehicleData.proximity goes from 0 to NUM_PROXIMITY_SENSOR_RANGES */
+        maxPossibleSpeed = m_maxMotorSpeed - (m_vehicleData.proximity * m_rangeFactor);
+
+        /* Ensure that the robot stops when an obstable is too near. */
+        if (NUM_PROXIMITY_SENSOR_RANGES == m_vehicleData.proximity)
+        {
+            maxPossibleSpeed = 0;
+        }
```
Comment (BlueAndi) [inline lib/ConvoyLeader/src/DrivingState.cpp:90]: calculate only in else part

File: lib/ConvoyLeader/src/DrivingState.cpp [hunk 7]
```diff
@@ -90,6 +102,7 @@ void DrivingState::exit()
 void DrivingState::setMaxMotorSpeed(int16_t maxSpeed)
 {
     m_maxMotorSpeed = maxSpeed;
+    m_rangeFactor   = m_maxMotorSpeed / NUM_PROXIMITY_SENSOR_RANGES;
```
Comment (BlueAndi) [inline lib/ConvoyLeader/src/DrivingState.cpp:105]: m_rangeFactor can be used just in place without member variable or?

File: lib/ConvoyLeader/src/DrivingState.cpp [hunk 8]
```diff
@@ -73,12 +73,24 @@ void DrivingState::process(StateMachine& sm)
...
     {
+        int16_t maxPossibleSpeed = m_maxMotorSpeed;
+
         /* TODO: Check limits. */
 
         /* TODO: Check follower feedback. Calculate platoon length and react accordingly */
 
+        /* Limit speed based on the proximity sensors. */
+        /* m_vehicleData.proximity goes from 0 to NUM_PROXIMITY_SENSOR_RANGES */
+        maxPossibleSpeed = m_maxMotorSpeed - (m_vehicleData.proximity * m_rangeFactor);
```
Comment (BlueAndi) [inline lib/ConvoyLeader/src/DrivingState.cpp:84]: Is there any processing time considered, e.g. from detection time to reaction time 10 ms or more? Or just a proof-of-concept?

**Review by BlueAndi — CHANGES_REQUESTED** [review]

File: lib/ConvoyFollower/src/App.cpp (hunk 2 above)
Comment (gabryelreyes) [reply lib/ConvoyFollower/src/App.cpp:497]: Error is unexpected error. Emergency is purposefully triggered.

Comment (BlueAndi) [reply lib/ConvoyFollower/src/App.cpp:489]: Shouldn't it be handled then different?

**Review by BlueAndi — CHANGES_REQUESTED** [review]

**Review by BlueAndi — APPROVED** [review]

**Pull request merged**